File: app/src/main/java/xyz/sleekstats/completist/model/MovieCredits.java
package xyz.sleekstats.completist.model;

import com.google.gson.annotations.SerializedName;

import java.util.HashSet;
import java.util.Set;

public class MovieCredits {

    @SerializedName("cast")
    private Set<FilmByPerson> cast;

    @SerializedName("crew")
    private Set<FilmByPerson> crew;

    private Set<FilmByPerson> both;

    public Set<FilmByPerson> getCrew() {
        return filterCrew(crew);
    }

    public Set<FilmByPerson> getCast() {
        return cast;
    }

    public Set<FilmByPerson> bothLists() {
        if(both == null) {
            Set<FilmByPerson> FilmByPersonSet = new HashSet<>(cast);
            FilmByPersonSet.addAll(getCrew());
            both = FilmByPersonSet;
        }
        return both;
    }

    private Set<FilmByPerson> filterCrew(Set<FilmByPerson> unfiltered) {
        Set<FilmByPerson> filteredSet = new HashSet<>();
        for (FilmByPerson film : unfiltered) {
            if (film.getJob().equals("Director") || film.getJob().equals("Writer") || film.getJob().equals("Screenplay")) {
                filteredSet.add(film);
            }
        }
        return filteredSet;
    }
}

File: app/src/main/java/xyz/sleekstats/completist/model/MovieCredits.kt
package xyz.sleekstats.completist.model

import com.google.gson.annotations.SerializedName
import java.util.*
import kotlin.collections.HashSet

class MovieCredits {
    @SerializedName("cast")
    val cast: Set<FilmByPerson> = emptySet()

    @SerializedName("crew")
    val crew: Set<FilmByPerson> = emptySet()
        get() = filterCrew(field)

    private var both: Set<FilmByPerson>? = null

    fun bothLists(): Set<FilmByPerson>? {
        if (both == null) {
            val filmByPersonSet: MutableSet<FilmByPerson> = HashSet(cast)
            filmByPersonSet.addAll(crew)
            both = filmByPersonSet
        }
        return both
    }

    private fun filterCrew(unfiltered: Set<FilmByPerson>): Set<FilmByPerson> = unfiltered.filterTo(HashSet()) { filteredCrewJobs.contains(it.job) }

    companion object {
        private val filteredCrewJobs = arrayOf("Director", "Writer", "Screenplay")
    }
}